### DISCUSSION TITLE
Copy mode: how can I copy a line with a single keystroke?

### DISCUSSION THREAD

**Discussion by rgwood** [wezterm/wezterm#5384] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

wezterm 20240203-110809-5046fc22

### Ask your question!

Hi! I am trying to tweak my keyboard shortcuts for copy mode. I would really like to have a single shortcut to copy the entire line the cursor is on, but I can't seem to make the `Multiple` action do that.

Here's an example from my `wezterm.lua` with number keys bound to actions:

```lua
local wezterm = require 'wezterm'
local act = wezterm.action
local copy_mode = wezterm.gui.default_key_tables().copy_mode

-- 1: set selection mode to line
table.insert(copy_mode, {
  key = '1',
  mods = 'NONE',
  action = act.CopyMode { SetSelectionMode = 'Line' },
})

-- 2: copy selection
table.insert(copy_mode, {
  key = '2',
  mods = 'NONE',
  action = act.CopyTo 'ClipboardAndPrimarySelection',
})

-- 3: set selection mode to line AND copy selection
table.insert(copy_mode, {
  key = '3',
  mods = 'NONE',
  action = act.Multiple {
    act.CopyMode { SetSelectionMode = 'Line' },
    act.CopyTo 'ClipboardAndPrimarySelection',
  }
})
```

If I press 1 then 2 in Copy Mode, it works as expected and the line the cursor is on is copied to the clipboard.

If I press 3, the entire line is selected but _not_ copied to the clipboard.

Anyone know what I'm doing wrong (or have another idea for how to achieve this)?

**Comment by rgwood** [comment]
Think there might be a bug in this area. I'm having more luck with an `action_callback`:

```lua
table.insert(copy_mode, {
  key = 'c',
  mods = 'CTRL|SHIFT',
  action = act.Multiple {
    act.CopyMode { SetSelectionMode = 'Line' },
    wezterm.action_callback(function(win, pane)
      local selected_text = win:get_selection_text_for_pane(pane)
      win:copy_to_clipboard(selected_text)
      win:perform_action(act.ClearSelection, pane)
    end),
}
})
```

However, this seems to work inconsistently on Linux (doesn't work the first time I try it after opening a new tab). Seems OK on Windows (same WezTerm version).

**Reply by wez** [reply]
hmm, this sounds like a timing issue. Under the covers, the declarative actions have an interface that is *synchronous* because they have to confirm to the keyboard handling layer above it (which is driven synchronously by the host OS) when they took action, but some actions have *asynchronous* completion, particularly those that work on the GUI layer.  Operating on the clipboard requires IPC, and that is exposed in its "API" surface on Linux, but hidden away in the macOS and Windows APIs.

When using `act.Multiple`, the listed actions are all initiated in the order you specify, but they are not guaranteed to complete before the next action in the list is started.

action_callback allows you to wait for function calls to complete, so you get better scheduling, but note that `win:perform_action` calls through to the same action dispatcher with the same limitations as mentioned above.

What I think you may want to try is using action_callback together with https://wezfurlong.org/wezterm/config/lua/wezterm/sleep_ms.html to insert a delay at the appropriate spot.  It's a bit gross.

**Reply by rgwood** [reply]
Thank you for the insight into how this works! I've switched over to doing everything in `action_callback` and that seems to do the trick on Linux+Windows with no `sleep_ms()` needed 🤞 

```lua
-- copy the entire line
table.insert(copy_mode, {
  key = 'c',
  mods = 'CTRL|SHIFT',
  action = wezterm.action_callback(function(win, pane)
      win:perform_action(act.CopyMode { SetSelectionMode = 'Line' }, pane)
      local selected_text = win:get_selection_text_for_pane(pane)
      -- optional: trim text here
      win:copy_to_clipboard(selected_text)
      win:perform_action(act.ClearSelection, pane)
      win:perform_action(act.CopyMode 'Close', pane)
    end),
})
```
